lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryFileSystemCache.java

```diff
@@ -0,0 +1,189 @@
+import com.google.common.cache.Weigher;
+import com.google.inject.Inject;
+import io.airlift.slice.Slice;
+import io.airlift.units.DataSize;
+import io.airlift.units.Duration;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.Tracer;
+import io.trino.cache.EvictableCacheBuilder;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoInput;
+import io.trino.filesystem.TrinoInputFile;
+import io.trino.filesystem.TrinoInputStream;
+import io.trino.filesystem.cache.TrinoFileSystemCache;
+import org.weakref.jmx.Managed;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.NoSuchFileException;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.slice.SizeOf.estimatedSizeOf;
+import static io.airlift.units.DataSize.Unit.GIGABYTE;
+import static io.trino.filesystem.tracing.CacheSystemAttributes.CACHE_FILE_LOCATION;
+import static io.trino.filesystem.tracing.CacheSystemAttributes.CACHE_FILE_READ_SIZE;
+import static io.trino.filesystem.tracing.CacheSystemAttributes.CACHE_KEY;
+import static io.trino.filesystem.tracing.Tracing.withTracing;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public final class MemoryFileSystemCache
+        implements TrinoFileSystemCache
+{
+    private final Tracer tracer;
+    private final int maxContentLengthBytes;
+    private final Cache<CacheKey, Slice> cache;
+
+    @Inject
+    public MemoryFileSystemCache(Tracer tracer, MemoryFileSystemCacheConfig config)
+    {
+        this(tracer, config.getCacheTTL(), config.getMaxSize(), config.getMaxContentLength());
+    }
+
+    public MemoryFileSystemCache(Tracer tracer, Duration expireAfterWrite, DataSize maxSize, DataSize maxContentLength)
+    {
+        this.tracer = requireNonNull(tracer, "tracer is null");
+        checkArgument(maxContentLength.compareTo(DataSize.of(1, GIGABYTE)) <= 0, "maxContentLength must be less than or equal to 1GB");
+        this.cache = EvictableCacheBuilder.newBuilder()
+                .maximumWeight(maxSize.toBytes())
+                .weigher((Weigher<CacheKey, Slice>) (key, value) -> toIntExact(estimatedSizeOf(key.providedKey()) + Integer.BYTES + value.getRetainedSize()))
+                .expireAfterWrite(expireAfterWrite.toMillis(), TimeUnit.MILLISECONDS)
+                .shareNothingWhenDisabled()
+                .recordStats()
+                .build();
+        this.maxContentLengthBytes = toIntExact(maxContentLength.toBytes());
+    }
+
+    @Override
+    public TrinoInput cacheInput(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        long fileSize = delegate.length();
```

Comment:
Will this always make a metadata request to object store, regardless of caching status? At least for Delta. With Iceberg, the file lengths for the various file types seems to be part of the spec for most (all?) of the file types.

Reply:
I've made changes to add this caching feature only to iceberg for now.
With iceberg we will make only one HEAD call for snapshot file (which would be there without caching as well), the rest of the metadata files get their length from iceberg metadata rather than remote filesystem call.
I could potentially remove this one call as well, as all iceberg metadata is immutable, but the presence of the escape hatch `use_file_size_from_metadata` makes me want to play it safe and keep the one metadata call.

Reply:
This also happens to match the way iceberg manifest cache is implemented at `org.apache.iceberg.io.ContentCache#tryCache(org.apache.iceberg.io.InputFile)`

Reply:
This option was added to resolve https://github.com/trinodb/trino/issues/6369 for when the file size metadata was incorrect or missing. Coordinator caching and `use_file_size_from_metadata` look like orthogonal features. We only run into trouble with caching if someone changes a metadata file, but the Iceberg spec explicitly calls out that data and metadata files are immutable until they are deleted: https://iceberg.apache.org/spec/#file-system-operations

It's unfortunate if we need to sequentially make one HEAD call for each Iceberg table in a query, even for repeated queries.

Reply:
Yeah `use_file_size_from_metadata` is actually unrelated, it's the test for it which manually deletes file and messes with the cache that caused me to think it's related.

Reply:
I've made changes to cache file length as well now, it required small updates to TrinoFileSystemCache